dev/tools/localization/gen_l10n.dart

```diff
@@ -302,184 +314,230 @@ bool _isValidGetterAndMethodName(String name) {
+  /// AppLocalizations will be used for localized message lookups.
+  ///
+  /// The class name is specified with the [initialize] method.
+  String className;
 
+  /// The list of all arb files in [l10nDirectory].
   final List<String> arbFilenames = <String>[];
+
+  /// The supported language codes as found in the arb files located in
+  /// [l10nDirectory].
   final Set<String> supportedLanguageCodes = <String>{};
-  final Set<LocaleInfo> supportedLocales = <LocaleInfo>{};
 
-  for (FileSystemEntity entity in l10nDirectory.listSync().toList()..sort(sortFilesByPath)) {
-    final String entityPath = entity.path;
-
-    if (FileSystemEntity.isFileSync(entityPath)) {
-      final RegExp arbFilenameRE = RegExp(r'(\w+)\.arb$');
-      if (arbFilenameRE.hasMatch(entityPath)) {
-        final File arbFile = File(entityPath);
-        final Map<String, dynamic> arbContents = json.decode(arbFile.readAsStringSync());
-        String localeString = arbContents['@@locale'];
-
-        if (localeString == null) {
-          final RegExp arbFilenameLocaleRE = RegExp(r'^[^_]*_(\w+)\.arb$');
-          final RegExpMatch arbFileMatch = arbFilenameLocaleRE.firstMatch(entityPath);
-          if (arbFileMatch == null) {
-            exitWithError(
-              "The following .arb file's locale could not be determined: \n"
-              '$entityPath \n'
-              "Make sure that the locale is specified in the '@@locale' "
-              'property or as part of the filename (ie. file_en.arb)'
-            );
-          }
+  /// The supported locales as found in the arb files located in
+  /// [l10nDirectory].
+  final Set<LocaleInfo> supportedLocales = <LocaleInfo>{};
 
-          localeString = arbFilenameLocaleRE.firstMatch(entityPath)[1];
-        }
+  /// The class methods that will be generated in the localizations class
+  /// based on messages found in the template arb file.
+  final List<String> classMethods = <String>[];
 
-        arbFilenames.add(entityPath);
-        final LocaleInfo localeInfo = LocaleInfo.fromString(localeString);
-        if (supportedLocales.contains(localeInfo))
-          exitWithError(
-            'Multiple arb files with the same locale detected. \n'
-            'Ensure that there is exactly one arb file for each locale.'
-          );
-        supportedLocales.add(localeInfo);
-        supportedLanguageCodes.add('\'${localeInfo.languageCode}\'');
-      }
-    }
+  /// Initializes [l10nDirectory], [templateArbFile], [outputFile] and [className].
+  void initialize({
+    String l10nDirectoryPath,
+    String templateArbFileName,
+    String outputFileString,
+    String classNameString,
+  }) {
+    setL10nDirectory(l10nDirectoryPath);
+    setTemplateArbFile(templateArbFileName);
+    setOutputFile(outputFileString);
+    setClassName(classNameString);
   }
 
-  final List<String> classMethods = <String>[];
+  /// Sets the reference [Directory] for [l10nDirectory].
+  void setL10nDirectory(String arbPathString) {
```

Comment:
You can make this private or visible for testing. Also, instead of `setMethodName`, use a setter:

```
set l10Directory(String absPath) {
 ...
}
```